api/group_creation/group_manager.py

```diff
@@ -0,0 +1,150 @@
+                TeacherStatus.TEACHING_ACCEPTING_MORE,
+            )
+        )
+
+    @staticmethod
+    def create_group(teacher_id: int) -> Group | None:
+        for group_candidate in GroupManager._iterate_group_candidates(teacher_id):
+            if group_candidate.is_acceptable():
+                group = Group(
+                    language_and_level=group_candidate.language_and_level,
+                    communication_language_mode=group_candidate.communication_language_mode,
+                    lesson_duration_in_minutes=DEFAULT_LESSON_DURATION_MIN,
+                    monday=group_candidate.monday,
+                    tuesday=group_candidate.tuesday,
+                    wednesday=group_candidate.wednesday,
+                    thursday=group_candidate.thursday,
+                    friday=group_candidate.friday,
+                    saturday=group_candidate.saturday,
+                    sunday=group_candidate.sunday,
+                )
+
+                group_creation_timestamp = timezone.now()
+                StatusSetter.set_status(
+                    obj=group, status=GroupStatus.PENDING, status_since=group_creation_timestamp
+                )
+                group.save()
+                group.teachers.add(group_candidate.teacher)
+                group.students.set(group_candidate.students)
+
+                # QQ: do we update teacher/student status automatically,
+                # or wait for coordinator action?
+                next_teacher_status = (
+                    TeacherStatus.TEACHING_ANOTHER_GROUP_OFFERED
+                    if group_candidate.teacher.status == TeacherStatus.TEACHING_ACCEPTING_MORE
+                    else TeacherStatus.GROUP_OFFERED
+                )
+                StatusSetter.set_status(
+                    obj=group_candidate.teacher,
+                    status=next_teacher_status,
+                    status_since=group_creation_timestamp,
+                )
+                GroupManager._create_log_events(group)
+                return group
+        # No suitable groups found
+        return None
+
+    @staticmethod
+    def _get_allowed_group_size(age_range: AgeRange) -> GroupSizeRestriction:
+        if MAX_AGE_TEEN_GROUP < age_range.age_from <= age_range.age_to:
+            return GroupManager.GroupSizeRestriction(min_number=5, max_number=10)
+        if MAX_AGE_KIDS_GROUP <= age_range.age_from <= age_range.age_to <= MAX_AGE_TEEN_GROUP:
+            return GroupManager.GroupSizeRestriction(min_number=2, max_number=8)
+        if 0 <= age_range.age_from <= age_range.age_to <= MAX_AGE_KIDS_GROUP:
+            return GroupManager.GroupSizeRestriction(min_number=2, max_number=6)
+        # If age range does not fall into expected ranges, fail early
+        raise ValueError(f"Group age range is inconsistent with boundaries: {age_range}")
+
+    @staticmethod
+    def _iterate_group_candidates(teacher_id: int) -> Iterable[GroupCandidate]:
```

Comment:
```suggestion
    def _iterate_group_candidates(teacher_id: int) -> Iterator[GroupCandidate]:
```
see SO discussion [here](https://stackoverflow.com/a/57916822).